### PR TITLE
APEXMALHAR-2303 Added S3RecordReaderModule for reading records line by line

### PR DESCRIPTION
**Pull request by ajaygit158-zz**
@chaithu14 @yogidevendra Kindly review

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java [hunk 1]
```diff
@@ -216,20 +217,45 @@ public FileBlockMetadata(String filePath, long blockId, long offset, long length
...
+      this.filePath = filePath;
+      this.fileLength = fileLength;
+    }
+
     public String getFilePath()
     {
       return filePath;
     }
 
+    public long getFileLength()
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java:249]: Add java docs

File: library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java [hunk 2]
```diff
@@ -216,20 +217,45 @@ public FileBlockMetadata(String filePath, long blockId, long offset, long length
...
+    }
+
     public FileBlockMetadata newInstance(@NotNull String filePath)
     {
       Preconditions.checkNotNull(filePath);
       return new FileBlockMetadata(filePath);
     }
+
+    public FileBlockMetadata newInstance(@NotNull String filePath, long fileLength)
+    {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java:265]: Why do we need this?

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java [hunk 3]
```diff
@@ -170,9 +177,41 @@ public void initialize(STREAM stream, BlockMetadata blockMetadata, boolean conse
...
       super.initialize(stream, blockMetadata, consecutiveBlock);
     }
 
+    /**
+     * Reads bytes from the stream starting from the offset into the buffer
+     *
+     * @param bytesFromCurrentOffset
+     *          bytes read till now from current block
+     * @param endByte
+     *          byte upto which the data is to be read
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java]: endByte is not a argument to readData. Why did you added endByte in javadocs ?

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+
+  /**
+   * Extracts the secretAccessKey from the given uri
+   *
+   * @param s3uri
+   *          given s3uri
+   * @return the secretAccessKey
+   */
+  protected static String extractSecretAccessKey(String s3uri)
+  {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: I think you can reuse extract\* API's from S3BlockReader.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+ * Typically, this operator will be connected to output of FileSplitterInput to
+ * read records in parallel.
+ */
+@Evolving
+public class S3RecordReader extends FSSliceReader
+{
+  private transient AmazonS3 s3Client;
+  private String bucketName;
+  private String accessKey;
+  private String secretAccessKey;
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Please add "endPoint" configuration. This is also needed for s3Client.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+ *
+ * Typically, this operator will be connected to output of FileSplitterInput to
+ * read records in parallel.
+ */
+@Evolving
+public class S3RecordReader extends FSSliceReader
+{
+  private transient AmazonS3 s3Client;
+  private String bucketName;
+  private String accessKey;
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Please add @NotNull to the operator properties.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+
+    /**
+     * Set the AmazonS3 service
+     *
+     * @param s3Client
+     *          given s3Client
+     */
+    public void setS3Client(AmazonS3 s3Client)
+    {
+      this.s3Client = s3Client;
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Please add @NotNull and use Preconditions.checkNotNull().

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+  }
+
+  /**
+   * Criteria for record split
+   *
+   * @param mode
+   *          Mode
+   */
+  public void setMode(RECORD_READER_MODE mode)
+  {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Add the setter method for mode as void setMode(String mode) and convert the string to RECORD_READER_MODE. So that, it's easy to configure this from properties file.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+    protected int readData(final long bytesFromCurrentOffset, final int bytesToFetch) throws IOException
+    {
+      GetObjectRequest rangeObjectRequest = new GetObjectRequest(bucketName, filePath);
+      rangeObjectRequest.setRange(offset + bytesFromCurrentOffset, offset + bytesFromCurrentOffset + bytesToFetch - 1);
+      S3Object objectPortion = s3Client.getObject(rangeObjectRequest);
+      S3ObjectInputStream wrappedStream = objectPortion.getObjectContent();
+      byte[] buffer = ByteStreams.toByteArray(wrappedStream);
+      wrappedStream.close();
+      this.setBuffer(buffer);
+      int bufferLength = buffer.length;
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Instead of setBuffer(),use the same buffer which is in super class.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReaderModule.java [hunk 10]
```diff
@@ -0,0 +1,367 @@
+/**
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReaderModule.java]: Move the new classes to org.apache.\* package.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+      /*
+       * Initialize the bufferSize and overflowBufferSize
+       */
+      int bufferSize = Long.valueOf(blockMetadata.getLength() - blockMetadata.getOffset()).intValue();
+      ((S3DelimitedRecordReaderContext)readerContext).setBufferSize(bufferSize);
+      if (overflowBufferSize > bufferSize) {
+        ((S3DelimitedRecordReaderContext)readerContext).setOverflowBufferSize(bufferSize);
+      } else {
+        ((S3DelimitedRecordReaderContext)readerContext).setOverflowBufferSize(overflowBufferSize);
+      }
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: I think you need to set the bufferSize, overflowBuffersize before initialize().

File: library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java (hunk 2 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java:265]: We had newInstance() method with filePath. Since we have added the new parameter, I thought of creating this method as well to support both filePath and fileLength

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java (hunk 11 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: This can actually be used before or after. It will not have any undesirable effect. In case of S3, we dont use the bufferSize to allocate memory to buffer.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java (hunk 9 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: I have changed this, however, I had to make it protected in the base class. I believe this will not be an issue.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReaderModule.java (hunk 10 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReaderModule.java]: Done!

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java (hunk 8 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Done

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java (hunk 4 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Done

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java (hunk 5 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Done

File: library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java (hunk 6 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java]: Done

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java (hunk 3 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java]: Had used this variable earlier. Corrected the issue

File: library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java (hunk 1 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/block/BlockMetadata.java:249]: Done

**Comment by ajaygit158-zz** [comment]
Incorporated @chaithu14 's review comments. 
@chaithu14 @yogidevendra Kindly review.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+ * 6. readersCount: count of readers to read input file<br/>
+ * 7. sequentialFileRead: if true, then each reader partition will read
+ * different file. <br/>
+ * instead of reading different offsets of the same file. <br/>
+ * (File level parallelism instead of block level parallelism)<br/>
+ * 8. blocksThreshold: number of blocks emitted per window
+ */
+@org.apache.hadoop.classification.InterfaceStability.Evolving
+public class S3RecordReaderModule implements Module
+{
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Extend this module from FSRecordReaderModule and override the createFileSplitter and createRecordReader methods.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java [hunk 13]
```diff
@@ -0,0 +1,588 @@
...
+ * This operator can be used for reading records/tuples from S3 in parallel
+ * (without ordering guarantees between tuples). Records can be delimited (e.g.
+ * newline) or fixed width records. Output tuples are byte[].
+ *
+ * Typically, this operator will be connected to output of FileSplitterInput to
+ * read records in parallel.
+ */
+@Evolving
+public class S3RecordReader extends FSSliceReader
+{
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: I seen most of the common functionality between S3RecordReader and FSRecordReader. So, please reuse the FSRecordReader.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java [hunk 14]
```diff
@@ -52,12 +55,12 @@
   /**
    * Criteria for record split
    */
-  private RECORD_READER_MODE mode = RECORD_READER_MODE.DELIMITED_RECORD;
+  protected RECORD_READER_MODE mode = RECORD_READER_MODE.DELIMITED_RECORD;
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java]: why protected ?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java [hunk 15]
```diff
@@ -52,12 +55,12 @@
...
 
   /**
    * Length for fixed width record
    */
-  private int recordLength;
+  protected int recordLength;
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java:64]: Why did you change it to protected ?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReaderModule.java [hunk 16]
```diff
@@ -370,24 +371,25 @@ public void setRepartitionCheckInterval(long repartitionCheckInterval)
...
    *
-   * @param mode
-   *          Mode
+   * @return mode
    */
-  public void setMode(RECORD_READER_MODE mode)
+  public String getMode()
   {
-    this.mode = mode;
+    return mode.toString();
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReaderModule.java]: Change to:  
    return mode; ?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.apex.malhar.lib.fs;
+
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: Can we move this to package: org.apache.apex.malhar.lib.io.fs.s3 ?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java [hunk 18]
```diff
@@ -0,0 +1,541 @@
...
+    // File path would be the path after bucket name.
+    // Check if the file path starts with "/"
+    if (filePath.startsWith("/")) {
+      filePath = filePath.substring(1);
+    }
+
+    if (mode == RECORD_READER_MODE.FIXED_WIDTH_RECORD) {
+      ((S3FixedWidthRecordReaderContext)readerContext).setFilePath(filePath);
+      ((S3FixedWidthRecordReaderContext)readerContext).setFileLength(block.getFileLength());
+    } else {
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: I think it will be good to verify the mode type is DELIMITED_RECORD.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java [hunk 19]
```diff
@@ -0,0 +1,541 @@
...
+  public void setSecretAccessKey(String secretAccessKey)
+  {
+    this.secretAccessKey = secretAccessKey;
+  }
+
+  /**
+   * S3 endpoint
+   *
+   * @param s3
+   *          endpoint
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: I think there is no parameter "s3" in setter method. Please verify and update it.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.apex.malhar.lib.fs;
+
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Can we move this to package: org.apache.apex.malhar.lib.io.fs.s3 ?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * This module is used for reading records/tuples from S3. Records can be read
+ * in parallel using multiple partitions of record reader operator. (Ordering is
+ * not guaranteed when records are read in parallel)
+ *
+ * Input S3 directory is scanned at specified interval to poll for new data.
+ *
+ * The module reads data in parallel, following parameters can be configured
+ * <br/>
+ * 1. files: list of file(s)/directories to read<br/>
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Please describe how s3 files be specified.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+ * 2. filePatternRegularExp: Files with names matching given regex will be read
+ * <br/>
+ * 3. scanIntervalMillis: interval between two scans to discover new files in
+ * input directory<br/>
+ * 4. recursive: if true, scan input directories recursively<br/>
+ * 5. blockSize: block size used to read input blocks of file<br/>
+ * 6. readersCount: count of readers to read input file<br/>
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Please verify the configuration parameters under description.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+   */
+  @Override
+  public S3RecordReader createRecordReader()
+  {
+    S3RecordReader s3RecordReader = new S3RecordReader();
+    s3RecordReader.setBucketName(S3BlockReader.extractBucket(getFiles()));
+    s3RecordReader.setAccessKey(S3BlockReader.extractAccessKey(getFiles()));
+    s3RecordReader.setSecretAccessKey(S3BlockReader.extractSecretAccessKey(getFiles()));
+    s3RecordReader.setEndPoint(s3EndPoint);
+    s3RecordReader.setMode(this.getMode().toString());
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Why do you converting mode to String ?

File: library/src/test/java/org/apache/apex/malhar/lib/fs/S3DelimitedRecordReaderTest.java [hunk 24]
```diff
@@ -0,0 +1,285 @@
...
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DAG;
+import com.datatorrent.lib.helper.OperatorContextTestHelper;
+import com.datatorrent.lib.io.block.AbstractFSBlockReader;
+import com.datatorrent.lib.io.block.BlockMetadata;
+import com.datatorrent.lib.testbench.CollectorTestSink;
+import com.datatorrent.netlet.util.Slice;
+
+@Ignore
+public class S3DelimitedRecordReaderTest
```
Comment (chaithu14) [inline library/src/test/java/org/apache/apex/malhar/lib/fs/S3DelimitedRecordReaderTest.java]: Instead of Ignore, I think we can remove these tests.

File: library/src/test/java/org/apache/apex/malhar/lib/fs/S3RecordReaderMockTest.java [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+    {
+      if (mode == RECORD_READER_MODE.DELIMITED_RECORD) {
+        ((S3DelimitedRecordReaderContextMock)readerContext).setFilePath(block.getFilePath());
+        ((S3DelimitedRecordReaderContextMock)readerContext).setFileLength(block.getFileLength());
+      } else {
+        ((S3FixedWidthRecordReaderContextMock)readerContext).setFilePath(block.getFilePath());
+        ((S3FixedWidthRecordReaderContextMock)readerContext).setFileLength(block.getFileLength());
+      }
+      return fs.open(new Path(block.getFilePath()));
+    }
```
Comment (chaithu14) [inline library/src/test/java/org/apache/apex/malhar/lib/fs/S3RecordReaderMockTest.java]: Change to :
  super.setupStream(block);
  return fs.open(new Path(block.getFilePath()));

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java (hunk 22 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Add endpoint

File: library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java (hunk 14 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java]: Changed

File: library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java (hunk 15 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReader.java:64]: Changed

File: library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReaderModule.java (hunk 16 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReaderModule.java]: Changed

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java (hunk 17 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: Done

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java (hunk 18 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: Changed

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java (hunk 19 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReader.java]: Updated

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java (hunk 20 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Updated

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java (hunk 21 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Added comments for specifying s3 files

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java (hunk 22 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Changed configuration parameters and added endpoint parameter

File: library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java (hunk 23 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/S3RecordReaderModule.java]: Changed

File: library/src/test/java/org/apache/apex/malhar/lib/fs/S3DelimitedRecordReaderTest.java (hunk 24 above)
Comment (ajaygit158-zz) [reply library/src/test/java/org/apache/apex/malhar/lib/fs/S3DelimitedRecordReaderTest.java]: I dont think we should remove these.  Though these tests dont run during travis build, they are good for developers to test locally on their system with S3. Also, if in future we move to a local proxy setup of S3, these test will be useful then.

File: library/src/test/java/org/apache/apex/malhar/lib/fs/S3RecordReaderMockTest.java (hunk 25 above)
Comment (ajaygit158-zz) [reply library/src/test/java/org/apache/apex/malhar/lib/fs/S3RecordReaderMockTest.java]: Changed

**Comment by ajaygit158-zz** [comment]
@chaithu14 Incorporated your comments.

**Comment by ajaygit158-zz** [comment]
@chaithu14 Incorporated review comments. Added Prefetching for overflowBlock to reduce calls to S3 server.

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java [hunk 26]
```diff
@@ -167,10 +174,54 @@ public LineReaderContext()
...
+    /**
+     * @param usedBytesFromOffset
+     *          number of bytes the pointer is ahead of the offset
+     * @return true if end of stream reached, false otherwise
+     */
+    protected boolean checkEndOfStream(final long usedBytesFromOffset)
+    {
+      if (!overflowBlockRead) {
+        return (offset - blockMetadata.getOffset() + usedBytesFromOffset < bufferSize);
+      } else {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java:210]: I think offset and blockMetadata.getOffset() are equal. If yes, then you can reduce the expression. Please correct it, if i am wrong.

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java [hunk 27]
```diff
@@ -167,10 +174,54 @@ public LineReaderContext()
...
+    }
+
+    /**
+     * Gives the number of bytes to be fetched from the stream
+     *
+     * @param readOverflowBlock
+     *          indicates whether we are reading main block or overflow block
+     * @return bytes to be fetched from stream
+     */
+    protected int calculateBytesToFetch(boolean readOverflowBlock)
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java]: Can you use overflowBlockRead instead of readOverflowBlock?

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java [hunk 28]
```diff
@@ -300,6 +372,15 @@ public Entity next() throws IOException
       }
       return null;
     }
+
+    @Override
+    protected int calculateBytesToFetch(boolean readOverflowBlock)
+    {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java:399]: Same here.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java [hunk 29]
```diff
@@ -0,0 +1,545 @@
...
+    }
+
+    RECORD_READER_MODE recordReaderMode = RECORD_READER_MODE.valueOf(this.getMode());
+    if (recordReaderMode == RECORD_READER_MODE.FIXED_WIDTH_RECORD) {
+      ((S3FixedWidthRecordReaderContext)readerContext).setFilePath(filePath);
+      ((S3FixedWidthRecordReaderContext)readerContext).setFileLength(block.getFileLength());
+    } else if (recordReaderMode == RECORD_READER_MODE.DELIMITED_RECORD) {
+      ((S3DelimitedRecordReaderContext)readerContext).setFilePath(filePath);
+      ((S3DelimitedRecordReaderContext)readerContext).setFileLength(block.getFileLength());
+    }
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: Can you move this block into initialize()?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java [hunk 30]
```diff
@@ -0,0 +1,545 @@
...
+    return null;
+  }
+
+  /**
+   * Initialize the reader context
+   */
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: Please correct the description.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+   * @return S3RecordReader instance
+   */
+  @Override
+  public S3RecordReader createRecordReader()
+  {
+    S3RecordReader s3RecordReader = new S3RecordReader();
+    s3RecordReader.setBucketName(S3BlockReader.extractBucket(getFiles()));
+    s3RecordReader.setAccessKey(S3BlockReader.extractAccessKey(getFiles()));
+    s3RecordReader.setSecretAccessKey(S3BlockReader.extractSecretAccessKey(getFiles()));
+    s3RecordReader.setEndPoint(s3EndPoint);
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java:87]: s3EndPoint is optional. Before setting, please check whether this is not null or not.

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java (hunk 26 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java:210]: As discussed offline, offset gets updated in next()

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java (hunk 27 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java]: Updated

File: library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java (hunk 28 above)
Comment (ajaygit158-zz) [reply library/src/main/java/com/datatorrent/lib/io/block/ReaderContext.java:399]: Updated

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java (hunk 29 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: Changed with additional type check for FileBlockMetadata

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java (hunk 30 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: The parent method also initializes the reader context, I added the additional functionality of initializing s3Client

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java (hunk 31 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java:87]: We already have the check for non null s3 endpoint when we are setting the endpoint to s3Client in S3RecordReader. Hence, avoiding this change as this will be an unnecessary check.

**Comment by ajaygit158-zz** [comment]
@chaithu14 Incorporated your review comments

**Comment by ajaygit158-zz** [comment]
@chaithu14 Incorporated your review comments. Kindly review.

**Comment by ajaygit158-zz** [comment]
@chaithu14 Incorporated your comments.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java [hunk 32]
```diff
@@ -0,0 +1,604 @@
...
+    }
+
+    /**
+     * Sets the length of the file to which the block belongs
+     *
+     * @param fileLength
+     *          length of the file to which the block belongs
+     */
+    public void setFileLength(long fileLength)
+    {
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: I think there is no use of setFilePath(), setFileLength(). Can you remove the setter methods for filePath and fileLength ?

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java [hunk 33]
```diff
@@ -0,0 +1,604 @@
...
+    }
+
+    /**
+     * Sets the S3RecordReaderParams object
+     *
+     * @param s3Params
+     *          S3RecordReaderParams object
+     */
+    protected void setS3Params(S3RecordReaderParams s3Params)
+    {
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: Same here.

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java [hunk 34]
```diff
@@ -0,0 +1,604 @@
...
+  }
+
+  /**
+   * Set the access key
+   *
+   * @param accessKey
+   *          given accessKey
+   */
+  public void setAccessKey(String accessKey)
+  {
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java:530]: Please add @NotNull and use Preconditions.checkNotNull().

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java [hunk 35]
```diff
@@ -0,0 +1,604 @@
...
+  }
+
+  /**
+   * Set the secretAccessKey
+   *
+   * @param secretAccessKey
+   *          secretAccessKey
+   */
+  public void setSecretAccessKey(String secretAccessKey)
+  {
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java:552]: Please add @NotNull and use Preconditions.checkNotNull().

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java [hunk 36]
```diff
@@ -0,0 +1,138 @@
...
+    s3RecordReader.setMode(this.getMode());
+    s3RecordReader.setRecordLength(this.getRecordLength());
+    if (overflowBlockSize != 0) {
+      s3RecordReader.setOverflowBufferSize(overflowBlockSize);
+      ;
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java]: Please remove line no: 92.

Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReaderModule.java]: Done

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java (hunk 35 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java:552]: Done

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java (hunk 34 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java:530]: Done

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java (hunk 33 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: Removed

File: library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java (hunk 32 above)
Comment (ajaygit158-zz) [reply library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3RecordReader.java]: Removed

**Comment by ajaygit158-zz** [comment]
@chaithu14 Incorporated your comments.

**Comment by chaithu14** [comment]
Looks good to me. I will wait for other comments from community. If not, i will merge it in 2 days.

**Comment by ajaygit158-zz** [comment]
Jenkins build is failing. Closing and repoening PR

**Comment by ajaygit158-zz** [comment]
Local Build is successful. Closing and reopening PR.
